persist and show containers state, last_state and related attributes

moolitayer opened the pull request:
persist & show state and last state for containers.

resolves #2663

![screenshot from 2015-07-30 16-30-28](https://cloud.githubusercontent.com/assets/3010449/8985410/ece923ae-36df-11e5-9e24-d547a218a7f5.png)

moolitayer commented:
fyi @simon3z @abonas 
@miq-bot add_label ui, providers/containers, enhancement

abonas commented on app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -341,8 +341,17 @@ def parse_container(container, pod_id)
       }
-
-      # TODO, state
+      Hash(container.state).each do |key, _val|
```
why is there a need to convert to hash and iterate here?

moolitayer replied:
convert to hash:
container.state !=  Hash(container.state) since container.state is of type Kubeclient::pod and container.state.each Is empty. is there a better way to extract data?.

As for the iteration, the returned data looks like so:
"state": {
  "waiting": {
    "reason": "Image: openshift/origin-deployer:v1.0.4 is ready, container is creating"
  }
I could just as well use first instead of each, but I need to get the key inside state due to the nesting.

abonas commented:
please check if tests should be added to the refresher tests.

moolitayer commented:
expanded with a bunch of state attributes, will include a more complete screenshot once I find a better way of testing this patch.

moolitayer commented:
An example with some non-null attributes:

![screenshot from 2015-08-05 16-37-23](https://cloud.githubusercontent.com/assets/3010449/9086935/606d2348-3b90-11e5-9c93-16daddc7ba05.png)

simon3z commented:
@moolitayer as mentioned in https://github.com/ManageIQ/manageiq/pull/3659#issuecomment-127035262 can you try to add refresher tests?

moolitayer commented:
added tests for:
started_at, last_state, last_exit_code, last_started_at, last_finished_at
(state was already tested)
changed last_state, signal and last_signal to integer according to types go

moolitayer commented on db/migrate/20150719123908_add_container_state.rb:
```diff
@@ -0,0 +1,17 @@
+class AddContainerState < ActiveRecord::Migration
+  def change
+    add_column :containers, :reason, :string
+    add_column :containers, :started_at, :string
```
@simon3z  Is string the correct data type here?

simon3z replied:
@moolitayer no, let's try to convert the datetime properly

simon3z commented on spec/models/ems_refresh/refreshers/kubernetes_refresher_spec.rb:
```diff
@@ -71,9 +71,16 @@ def assert_authentication
+      :last_state       => "terminated",
+      :last_exit_code   => 1,
+      :last_started_at   => "2015-07-29T15:48:42Z",
+      :last_finished_at => "2015-07-29T15:48:42Z",
```
These will be a mess next time we'll do a new recording :sob: cc @blomquisg @oschreib @abonas

oschreib replied:
:(
Well, I don't see any decent workaround that.
We could either record (and change the recording to match these, I know, ugly), or change it on any new recording.
Hopefully, as we're on V1 everywhere now, we won't have to re-record everything.

zeari replied:
> These will be a mess next time we'll do a new recording :sob:

@simon3z @oschreib These should be commented out like ```ems_ref```.  The overhead of changes for every recording is too much.

moolitayer commented:
I will just check

simon3z commented on app/models/ems_refresh/parsers/kubernetes.rb:
```diff
@@ -349,8 +349,29 @@ def parse_container(container, pod_id)
+          :state       => key,
+          :reason      => val['reason'],
+          :started_at  => val['startedAt'],
+          :finished_at => val['finishedAt'],
```
@moolitayer datetimes should be explicitly converted here (not implicitly by activerecord or the db). (Right? here `finishedAt` and all the others are strings).

simon3z commented on spec/models/ems_refresh/refreshers/kubernetes_refresher_spec.rb:
```diff
@@ -71,12 +71,20 @@ def assert_authentication
+    iso = %r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$"
+    @container.started_at.should match (iso)
+    @container.last_started_at.should match (iso)
+    @container.last_finished_at.should match (iso)
```
Then here you can just validate that they're a date `kind_of?`.

blomquisg commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -362,8 +362,33 @@ def parse_container(container, pod_id)
       }
-
-      # TODO, state
+      Hash(container.state).each do |key, val|
```
How many `container.state`s are there?  This code seems to walk through several key/value pairs and calls `h.merge!` each time with the same set of keys.  Which would simply overwrite any values from previous iterations in the loop.  So, it looks like calling `Hash(container.state).last` has the same effect here and possibly saves the CPU time of looping through multiple `container.state`s.

If there's only meant to be 0 or 1 `container.state`, then I think `.last` is still effective here with a appropriately placed `nil` check.

However, what if there is nothing in `container.state`?  Should the keys set in this block be set to some kind of initial value?

In all, using `.each` here is quite confusing for a lot of reasons.  I suggest either switching to `.last` with a `nil` check, or re-examining the body of this block to do something more useful than just implementing a "last write wins" model.

matthewd replied:
Consider `(state, state_info), = container.state.to_a` to unpack the single-unknown-key hash.

moolitayer replied:
Defiantly confusing. Here is what I need to do, I have one of

```
"state": {
  "waiting": {
    "reason": "Image: openshift/origin-deployer:v1.0.4 is ready, container is creating"
  }
}

OR

"state": {
  "running": {
    "StartedAt": somedate
  }
}

OR

"state": {
  "terminated": {
    "exitcode": ...,
    "signal": ...,
    "message": ...,
    ... bunch of other attributes
  }
}
```
One of those (terminated, running, waiting) must be present, some attributes inside state are optional.

what I take from above is that I should patch:

\- Hash(container.state).each do |key, val|
\+ Hash(container.state).first do |key, val|

anything different?

matthewd replied:
Yeah, that's what I thought. Do the thing that I said.

(Why are you calling `Hash()`?)

blomquisg replied:
Yeah, I agree with @matthewd here.  Convert `container.state` to an array to capture the `state` and `state_info`.  Then, you can parse the `state_info` as a simple hash.

You should probably add a comment to the `container.state.to_a` call to make it clear what you're parsing.  In fact, you might just want to pull this all out to a method that parses the container state, and add a comment there with these examples, so it's clear exactly what's getting parsed.

blomquisg commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -362,8 +362,33 @@ def parse_container(container, pod_id)
+          :message     => val['message'],
+        )
+      end
+      Hash(container.lastState).each do |key, val|
```
[Same](https://github.com/ManageIQ/manageiq/pull/3659/files#r37754702)

moolitayer commented:
@matthewd  
> (Why are you calling Hash()?)

since container.last_state is of type Kubeclient::Pod, otherwise it is empty

moolitayer commented:
addressed comments, verified

moolitayer commented:
@simon3z @abonas

simon3z commented:
LGTM :+1: @matthewd @blomquisg care to review/merge? Thanks!

blomquisg commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -362,8 +362,28 @@ def parse_container(container, pod_id)
+      h.merge! parse_container_state container.state
+    end
+
+    def parse_container_state(state_hash)
```
Maybe have this take an additional `prefix` argument:

```ruby
   def parse_container_state(state_hash, prefix = "")
     ...
     res["#{prefix}state".to_sym] = state
     %w(reason started_at finished_at exit_code signal message).each do |attr|
       res["#{prefix}#{attr}".to_sym] = state_info[attr.camelize(:lower)]
     end
     ...
   end
```

Then the caller could just do:

```ruby
  h.merge! parse_container_state(container.lastState, "last_")
  h.merge! parse_container_state(container.state)
```

The way it reads currently, there's a bit of extra parsing that gets in the middle of understanding what's really being parsed.

The key here is to make the parsing as clear as possible in as many places as possible.

Thoughts?

moolitayer replied:
Incidentally, I did just that. I'm trying to remember why I abandoned it and changed it to be this way and I'm pretty sure I couldn't get parse_container_state simple enough  for rubocop's taste that way ( 'assignment branch condition ...'). If you want me to give another go let me know, maybe you could help with rubocop this time...
10x

blomquisg replied:
Heh, if what I'm saying is at odds with Rubocop, forget it.  It's probably not worth battling the authorities over :)

blomquisg commented:
This looks quite a bit better.  Take a look at my last [comment](https://github.com/ManageIQ/manageiq/pull/3659#discussion_r38339101) and let me know if that makes sense.

moolitayer commented:
@blomquisg

> This looks quite a bit better. Take a look at my last comment and let me know if that makes sense.

addressed inline please respond.

blomquisg commented:
@moolitayer, hrm .. guessing something got merged before this and caused a merge conflict.  Take another look and I'll see if I can merge in the morning (ET).

moolitayer commented:
@blomquisg 
done.
looks like this conflicted with https://github.com/ManageIQ/manageiq/pull/4232

simon3z commented:
LGTM ￼:+1: @blomquisg review/merge? thanks!
